extension-manifest-v3/src/background/reporting.js

```diff
@@ -0,0 +1,138 @@
+  ) {
+    // Ghostery extension for Safari on MacOS (Desktop)
+    return {
+      ALLOWED_COUNTRY_CODES: ['us', 'de', 'fr'],
+      PATTERNS_URL:
+        'https://cdn2.ghostery.com/wtm-safari-desktop/patterns.json',
+      CHANNEL: 'safari-desktop',
+    };
+  }
+
+  console.warn(
+    'No matching config found. Falling back to patterns from Chrome Desktop.',
+  );
+  return {
+    ALLOWED_COUNTRY_CODES: ['us', 'de', 'fr'],
+    PATTERNS_URL: 'https://cdn2.ghostery.com/wtm-chrome-desktop/patterns.json',
+    CHANNEL: 'ghostery',
+  };
+}
+
+const COLLECTOR_DIRECT_URL = 'https://anonymous-communication.ghostery.net';
+const COLLECTOR_PROXY_URL = COLLECTOR_DIRECT_URL; // current we have no proxy configured
+
+const config = {
+  COLLECTOR_DIRECT_URL,
+  COLLECTOR_PROXY_URL,
+  CONFIG_URL: 'https://api.ghostery.net/api/v1/config',
+  ...platformSpecificSettings(),
+};
+console.debug('reporting config:', config);
+
+class Storage {
+  constructor(namespace) {
+    this.namespace = `wtm::v1::${namespace}::`;
+  }
+
+  async get(key) {
+    const prefixedKey = this.namespace + key;
+    const result = await chrome.storage.local.get(prefixedKey);
+    return result[prefixedKey];
+  }
+
+  async set(key, value) {
+    const prefixedKey = this.namespace + key;
+    return chrome.storage.local.set({ [prefixedKey]: value });
+  }
+}
+
+const communication = new AnonymousCommunication({
+  config,
+  storage: new Storage('communication'),
+});
+const reporting = new Reporting({
+  config,
+  storage: new Storage('reporting'),
+  communication,
+});
+
+observe('terms', (terms) => {
+  if (terms) {
+    reporting.init().catch((e) => {
+      console.warn(
+        'Failed to initialize reporting. Leaving the module disabled and continue.',
+        e,
+      );
+    });
+  } else {
+    reporting.unload();
+  }
+});
+
+chrome.webNavigation.onCommitted.addListener((details) => {
+  const { url, frameId } = details;
+  if (frameId !== 0 || url === 'about:blank' || url.startsWith('chrome://')) {
+    return;
+  }
+
+  // TODO: in private windows, we should also exit (how to get the information?)
+
+  (async () => {
+    try {
+      const jobRegistered = await reporting.analyzeUrl(url);
+      if (jobRegistered) {
+        // TODO: This part here is not robust:
+        // we should avoid timers in MV3 or at least assume that we the service
+        // worker will die (persisting the jobs and shift the scheduling
+        // responsibility into the reporting module itself could help)
+        const delay = ({ timeInMs }) => {
+          return new Promise((resolve) => setTimeout(resolve, timeInMs));
+        };
+        await delay({ timeInMs: 2000 + 3000 * Math.random() });
+
+        await reporting.processPendingJobs();
+      }
+    } catch (e) {
+      console.warn('Unexpected error in reporting module:', e);
+    }
+  })();
+});
+
+// for debugging service-workers (TODO: provide a way to control logging)
+self.ghostery = self.ghostery || {};
```

Comment:
Clean up debugging code.

Reply:
This is also intentional to have a way to interact with the objects.

Reply:
If you really need that for debugging, I would recommend using `globalThis` instead of `self`, as it is normalized access to the global object regardless of the context. 

However, it introduces a side-effect, and I am not still convinced if it should be run without any condition, or behind the function call.

Reply:
I changed it to globalThis.

Regarding side-effects, it extends a global called "ghostery". I would hope this should not impact other APIs. The use case is to inspect the state when you have a production build and cannot change the code. Or to trigger operations like forcing a pattern update. I'm not aware of a better mechanism then using globals, apart from building extra tooling into the extension and exposing them through the UI.

Reply:
Ok, good point with the argument about the production build. You've bought me :)